### PR TITLE
Use json tags for yang-serialization naming and allow derived types in stringlists

### PR DESCRIPTION
**Pull request by HRogge**
Instead of just guessing (by shortening/camelcase) the go fieldname of a struct, look for json tags that define a fitting name.

Allow types derived from string/float64 to be used in Stringlists, e.g.:
`type Specialtype float64`

### CODE REVIEW

File: nodeutil/reflect.go [hunk 1]
```diff
@@ -515,6 +515,13 @@ func (self Reflect) WriteFieldWithFieldName(fieldName string, m meta.Leafable, p
 	}
 
 	fieldVal := elemVal.FieldByName(fieldName)
+	if !fieldVal.IsValid() {
+		panic(fmt.Sprintf("Invalid property \"%s\" on %s", fieldName, elemVal.Type()))
+	}
+	if v == nil {
+		panic(fmt.Sprintf("No value given to set %s", m.Ident()))
+	}
+
```
Comment (dhubler) [inline nodeutil/reflect.go:521]: I should have left comment in original code, but this was after `f.When` intentionally so that When has a chance to work when reflection doesn't

File: nodeutil/reflect.go [hunk 2]
```diff
@@ -631,6 +630,37 @@ func (self Reflect) ReadFieldWithFieldName(fieldName string, m meta.Leafable, pt
 	return v, nil
 }
 
+func MetaNameToFieldNameExt(parent reflect.Value, in string) string {
```
Comment (dhubler) [inline nodeutil/reflect.go]: What does "Ext" mean here? Extended? Is there a name that is more specific to what role this function plays?

File: val/conv.go [hunk 3]
```diff
@@ -902,5 +902,24 @@ func toStringList(val interface{}) ([]string, error) {
...
+		l := make([]string, rv.Len())
+		for i := range l {
+			s, err := toString(rv.Index(i).Interface())
+			if err != nil {
+				return nil, err
+			}
+			l[i] = s
+		}
+		return l, nil
+	}
```
Comment (dhubler) [inline val/conv.go:895]: seems unrelated. Sometimes this is ok, but with a feature as interesting as tags, i'd rather that be in MR by itself.  Also need unit test for this.

**Review by dhubler — COMMENTED** [review]
Thanks for the MR!  While I originally didn't do tags originally because I didn't want to confuse folks w/too many options, Reflect really is turning out to be useful and I think making it more powerful would be prudent if we documented it well.  We'll need unit tests.

File: nodeutil/reflect.go (hunk 2 above)
Comment (HRogge) [reply nodeutil/reflect.go]: Extended... because the original function MetaNameToFieldName() is public, so it cannot be changed. I am open to other suggestions how to hook this is well.

File: val/conv.go (hunk 3 above)
Comment (HRogge) [reply val/conv.go:895]: I have to admit I was a bit lazy with this PR... I was solving a local issue and ended up with two things to solve it (after a LOT of experimentation). But its two commits, so it should be easy to split apart.

You are right about the unit tests... **sigh**

so split in two branches, try to add a test and then make two new PRs?

Comment (dhubler) [reply val/conv.go:895]: 1 MR and separate commits is fine.

File: nodeutil/reflect.go (hunk 2 above)
Comment (dhubler) [reply nodeutil/reflect.go]: how about just MetaNameToField because you are not nec. using field name as you could potentially be using tags

File: val/conv.go (hunk 3 above)
Comment (HRogge) [reply val/conv.go:895]: Is there already a testcase for the MetaNameToField mechanics?

Comment (HRogge) [reply val/conv.go:895]: found it... reflect_test.go TestMetaNametoFieldName... will see that I extend this

**Comment by HRogge** [comment]
I added some more testcases for both commits of the PR and cleaned up the flow of toString() and toStringList() at the same time.

File: nodeutil/reflect.go (hunk 2 above)
Comment (HRogge) [reply nodeutil/reflect.go]: Sure... I am not that good with function names, feel free to edit it when you merge.

Comment (HRogge) [reply nodeutil/reflect.go]: I renamed the function and added a comment to explain what its doing

File: nodeutil/reflect.go (hunk 1 above)
Comment (HRogge) [reply nodeutil/reflect.go:521]: Not sure there is much you can do if reflect is not working. The reason I moved the check was that a lot of my early experiments with When() functions threw a panic because many reflect functions don't work on nil or "invalid" reflect values.

Comment (dhubler) [reply nodeutil/reflect.go:521]: correct.  When reflection is  invalid and When.OnRead and When.OnWrite is still called then implementations would call/read methods/field directly w/o using reflection.  I do see why you thought this was a bug tho.  If we change it now, it would change valid (admit-tingly not intuitive) behavior.

Comment (HRogge) [reply nodeutil/reflect.go:521]: then we should at least add some warning comments that the input can be invalid/nil... so everyone knows that the functions have to check.

Comment (dhubler) [reply nodeutil/reflect.go:521]: agreed.

**Comment by HRogge** [comment]
ready to pull? If not, please tell me what I need to do.

**Comment by dhubler** [comment]
I looked it over.  I have one question, why `json` for a tag and not say `fc`?  While it may often be the same, I could see someone wanting to control these independently and not being able to.

**Comment by HRogge** [comment]
My reasoning was that the tag is to define the JSON serialization for freeconf, so reusing that json tag sounded okay. We could rename it to "fc-json" and use the "json" tag as a fallback if nothing else matches. The interesting question is if we need a "fc-xml" tag when the code finally supports xml too

**Comment by dhubler** [comment]
When you're reading and writing data from your business objects, you do not
know or care what the format is on the other side: JSON, XML, CBOR,
Protobuf.  All are possible and future ones inevitable.  So "fc" or maybe
"yang"
